mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java

```diff
@@ -0,0 +1,345 @@
+
+    public ConfBuilder(Configuration conf, String path, Class<? extends ReadSupport<?>> readSupportClass) {
+      this.conf = conf;
+      conf.set(TABLE_PATH, path);
+      conf.set(READ_SUPPORT, readSupportClass.getName());
+      Table table = getTable(conf);
+      conf.set(TABLE_SCHEMA, SchemaParser.toJson(table.schema()));
+    }
+
+    public ConfBuilder filterExpression(Expression expression) throws IOException {
+      conf.set(FILTER_EXPRESSION, SerializationUtil.serializeToBase64(expression));
+      return this;
+    }
+
+    public ConfBuilder project(Schema schema) {
+      conf.set(READ_SCHEMA, SchemaParser.toJson(schema));
+      return this;
+    }
+
+    //TODO: other options split-size, snapshotid etc..
+    public Configuration updatedConf() {
+      return conf;
+    }
+  }
+
+  private static final class IcebergRecordReader<T> extends RecordReader<Void, T> {
+    private TaskAttemptContext context;
+    private Iterator<FileScanTask> tasks;
+    private Iterator<T> currentIterator;
+    private T currentRow;
+    private Schema expectedSchema;
+    private Schema tableSchema;
+    private ReadSupport<T> readSupport;
+    private Closeable currentCloseable;
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext context) {
+      Configuration conf = context.getConfiguration();
+      CombinedScanTask task = ((IcebergSplit) split).task;
+      this.context = context;
+      this.tasks = task.files().iterator();
+      this.tableSchema = SchemaParser.fromJson(conf.get(TABLE_SCHEMA));
+      String readSchemaStr = conf.get(READ_SCHEMA);
+      if (readSchemaStr != null) {
+        this.expectedSchema = SchemaParser.fromJson(readSchemaStr);
+      }
+      this.readSupport = readSupport(conf);
+      this.currentIterator = open(tasks.next());
+    }
+
+    @Override
+    public boolean nextKeyValue() throws IOException {
+      while (true) {
+        if (currentIterator.hasNext()) {
+          currentRow = currentIterator.next();
+          return true;
+        } else if (tasks.hasNext()) {
+          currentCloseable.close();
+          currentIterator = open(tasks.next());
+        } else {
+          return false;
+        }
+      }
+    }
+
+    @Override
+    public Void getCurrentKey() {
+      return null;
+    }
+
+    @Override
+    public T getCurrentValue() {
+      return currentRow;
+    }
+
+    @Override
+    public float getProgress() {
+      return context.getProgress();
+    }
+
+    @Override
+    public void close() throws IOException {
+      currentCloseable.close();
+    }
+
+    private ReadSupport<T> readSupport(Configuration conf) {
+      String readSupportClassName = conf.get(READ_SUPPORT);
+      try {
+        return DynClasses
+            .builder()
+            .impl(readSupportClassName)
+            .<ReadSupport<T>>buildChecked()
+            .newInstance();
+      } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
```

Comment:
You might want to use `DynConstructors` and `build` instead of `buildChecked`. That will get you better error messages and some of this error handling for free.